pkg/util/fxutil/provide_comp.go

```diff
@@ -0,0 +1,381 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package fxutil
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+
+	compDef "github.com/DataDog/datadog-agent/comp/def"
+	"go.uber.org/fx"
+)
+
+var (
+	compOutType = reflect.TypeOf((*compDef.Out)(nil)).Elem()
+	compLcType  = reflect.TypeOf((*compDef.Lifecycle)(nil)).Elem()
+	fxInType    = reflect.TypeOf((*fx.In)(nil)).Elem()
+	fxOutType   = reflect.TypeOf((*fx.Out)(nil)).Elem()
+	fxLcType    = reflect.TypeOf((*fx.Lifecycle)(nil)).Elem()
+)
+
+// ProvideComponentConstructor takes as input a Component constructor function
+// that uses plain (non-fx aware) structs as its argument and return value, and
+// returns an fx.Provide'd Option that will properly include that Component
+// into the fx constructor graph.
+//
+// For example, given:
+//
+//	type Provides struct {
+//	    My MyComponent
+//	}
+//	type Requires struct {
+//	    Dep MyDependency
+//	}
+//	func NewComponent(reqs Requires) Provides { ... }
+//
+// then:
+//
+//	ProvideComponentConstructor(NewComponent)
+//
+// will create these anonymous types:
+//
+//	type FxAwareProvides struct {
+//	    fx.Out
+//	    My MyComponent
+//	}
+//	type FxAwareRequires struct {
+//	    fx.In
+//	    Dep MyDependency
+//	}
+//
+// and essentially make this happen:
+//
+//	fx.Provide(func(fxr FxAwareRequires) fxp FxAwareProvides {
+//	    r := makePlainReqs(fxr)
+//	    p := NewComponent(r)
+//	    return makeFxAwareProvides(p)
+//	})
```

Comment:
Do we need this part in the doc string for the function ?

Reply:
No, I can move it up higher, it doesn't need to be right here.

Reply:
You think it's too focused on implementation? I suppose that's true. I thought it made sense to explain what happens with the created anonymous fx types, but I can summarize instead of getting too detailed.